crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -1455,33 +1393,217 @@ impl AuthorityPerEpochStore {
         &self,
-        transactions: Vec<VerifiedSequencedConsensusTransaction>,
+        batch: &mut DBBatch,
+        mut transactions: Vec<VerifiedSequencedConsensusTransaction>,
         checkpoint_service: &Arc<C>,
         parent_sync_store: impl ParentSync,
-    ) -> SuiResult<Vec<VerifiedExecutableTransaction>> {
+    ) -> SuiResult<(
+        Vec<VerifiedExecutableTransaction>,
+        Option<parking_lot::RwLockWriteGuard<ReconfigState>>,
+    )> {
         let mut verified_certificates = Vec::new();
-        for tx in transactions {
+
+        // sort all end of publish messages to the end of the list
+        transactions.sort_by(|a, b| {
+            let a_is_end_of_publish = a.0.is_end_of_publish();
+            let b_is_end_of_publish = b.0.is_end_of_publish();
+
+            a_is_end_of_publish.cmp(&b_is_end_of_publish)
+        });
+
+        // partition transactions into two groups: end of publish messages and all other messages
+        let (end_of_publish_txns, other_txns): (Vec<_>, Vec<_>) = transactions
+            .into_iter()
+            .partition(|tx| tx.0.is_end_of_publish());
+
+        // get the current next versions for each shared object in the other_txns
+        let mut shared_input_next_versions = {
+            let unique_shared_input_objects = {
+                let mut shared_input_objects: Vec<_> = other_txns
+                    .iter()
+                    .filter_map(|tx| tx.0.as_shared_object_txn())
+                    .flat_map(|tx| {
+                        tx.transaction_data()
+                            .shared_input_objects()
+                            .into_iter()
+                            .map(|so| so.into_id_and_version())
+                    })
+                    .collect();
+
+                shared_input_objects.sort();
+                shared_input_objects.dedup();
+                shared_input_objects
+            };
+
+            self.get_or_init_next_object_versions(
+                unique_shared_input_objects.into_iter(),
+                &parent_sync_store,
+            )
+            .await?
+        };
+
+        for tx in &other_txns {
             if let Some(cert) = self
-                .process_consensus_transaction(tx, checkpoint_service, &parent_sync_store)
+                .process_consensus_transaction(
+                    batch,
+                    &mut shared_input_next_versions,
+                    tx,
+                    checkpoint_service,
+                )
                 .await?
             {
                 verified_certificates.push(cert);
             }
         }
-        Ok(verified_certificates)
+
+        batch.insert_batch(
+            &self.tables.next_shared_object_versions,
+            shared_input_next_versions.into_iter(),
+        )?;
+
+        let lock = self.process_end_of_publish_transactions(batch, &end_of_publish_txns)?;
+
+        for tx in other_txns.iter().chain(end_of_publish_txns.iter()) {
```

Comment:
We should do this after batch is written, otherwise notify_read might hang in some cases due to race condition